File: lab6-openmp/lab6-openmp_f90/omp_bug3.f90
! -*- Mode: Fortran90; -*-
!=================================================================

program BUG4
  !-----------------------------------------------------------------
  ! FILE: omp_bug4.f90
  ! DESCRIPTION:
  !   This very simple program causes a segmentation fault.
  ! AUTHOR: Blaise Barney  01/09/04
  ! Upgraded to F90, free-source form: Daniel R. Reynolds, 2/2009
  !-----------------------------------------------------------------
  !======= Inclusions ===========

  !======= Declarations =========
  implicit none

  integer :: NThreads, TID, i, j
  integer, parameter :: N=1048
  integer, external :: omp_get_num_threads, omp_get_thread_num
  real*8 :: A(N,N)

  !======= Internals ============

  ! Fork a team of threads with explicit variable scoping
  !$omp parallel shared(NThreads) private(i,j,TID,A)

  ! Obtain/print thread info
  TID = omp_get_thread_num()
  if (TID == 0) then
     NThreads = omp_get_num_threads()
     print *, 'Number of threads = ', NThreads
  end if
  print *, 'Thread',TID,' starting...'

  ! Each thread works on its own private copy of the array
  do i=1,N
     do j=1,N
        A(j,i) = 1.d0*(TID + i + j)
     end do
  end do

  ! For confirmation
  print *, 'Thread',TID,'done. Last element=',A(N,N)

  ! All threads join master thread and disband
  !$omp end parallel

end program BUG4
!=================================================================

File: lab6-openmp/lab6-openmp_cxx/omp_bug3.cpp
/* FILE: omp_bug3.cpp
   DESCRIPTION:
      This very simple program causes a segmentation fault.
   AUTHOR: Blaise Barney  01/09/04
   UPDATED: Daniel R. Reynolds (Updated to C++), 1/13/2013 */

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>
#define N 1048

int main (int argc, char *argv[]) {

  // local variables
  int nthreads, tid, i, j;
  double a[N][N];

  // Fork a team of threads with explicit variable scoping
# pragma omp parallel shared(nthreads) private(i,j,tid,a)
  {

    // Obtain/print thread info
    tid = omp_get_thread_num();
    if (tid == 0) {
      nthreads = omp_get_num_threads();
      printf("Number of threads = %i\n", nthreads);
    }
    printf("Thread %i starting...\n", tid);

    // Each thread works on its own private copy of the array
    for (i=0; i<N; i++)
      for (j=0; j<N; j++)
	a[i][j] = tid + i + j;

    // For confirmation
    printf("Thread %i done. Last element = %g\n", tid, a[N-1][N-1]);

  }  // All threads join master thread and disband

  return 0;
}  // end main

